design/func-version.md

```diff
@@ -0,0 +1,378 @@
+# Kanister Functions Version Management
+
+<!-- toc -->
+- [Problem Statement](#problem-statement)
+- [Proposed Solution](#proposed-solution)
+  - [Summary](#summary)
+  - [Assumptions And Constraints](#assumptions-and-constraints)
+  - [Function Versioning](#function-versioning)
+  - [Function Registration And Discovery](#function-registration-and-discovery)
+  - [User Experience](#user-experience)
+  - [Version Promotion And Deprecation](#version-promotion-and-deprecation)
+  - [Versions Discovery](#versions-discovery)
+- [Test Cases](#test-cases)
+<!-- /toc -->
+
+## Problem Statement
+
+As Kanister continues to grow and evolve to meet the needs of its community,
+it's inevitable that eventually, some of the existing built-in [Kanister
+Functions][0] will be updated with breaking changes, deprecated or even
+removed entirely.
+
+The lack of a proper strategy to manage and rollout such changes in a
+responsible backwards compatible manner can lead to costly operational overhead
+for Kanister users. In addition, it also becomes difficult for Kanister
+maintainers to assess the blast radius of the change, determine the rollout
+timing, provide timely support etc.
+
+This proposal outlines a version management scheme that can be used to manage
+the lifecycle of Kanister Functions as they evolve from inception to retirement.
+
+## Proposed Solution
+
+### Summary
+
+The proposed solution uses a version management scheme to communicate the
+stability, availability, promotion and deprecation of Kanister Functions.
+
+The design extends the existing implementation of the action-level
+[`PreferredVersion` mechanism][1] to the phase-level by enabling users to
+specify the desired  Kanister Function version within the blueprint. It also
+proposes changes to the relevant interfaces with code examples demonstrating the
+implementation of multiple versions of a Kanister Function.
+
+### Assumptions And Constraints
+
+* The proposed solution is compatible with existing downstream custom Kanister
+Function implementation, without any code change.
+
+### Function Versioning
+
+The proposed solution introduces a new Go interface named `VersionedFunc`. The
+`VersionedFunc` type embeds the `Func` type to include information about the
+multiple versions of implementation of the underlying Kanister Function.
+
+```go
+// VersionedFunc extends the Func type to include information about the multiple
+// versions of the underlying Kanister Function.
+type VersionedFunc interface {
+  // Func is the v0.0.0 version of the function. For existing functions, this
+  // maps to their existing `Func` implementation.
+  Func
+
+  // StableVersion returns the stable version of the function.
+  StableVersion() string
+
+  // Versions maps the different versions of the function to their respective
+  // implementations.
+  Versions() map[string]Func
```

Comment:
We obviously need this, but I am not sure this function should be part of this interface. More comments below here this is implemented.

Reply:
this name is a bit confusing :wink:

Reply:
How so? It's meant to be a getter, without adding the `Get` prefix, per Go's convention.